src/bootloaders/shim-systemd.c

```diff
@@ -0,0 +1,313 @@
+        "systemd-boot" EFI_SUFFIX
+#define DST_DIR "/" KERNEL_NAMESPACE
+#define SHIM_DST                                                                                   \
+        DST_DIR                                                                                    \
+        "/"                                                                                        \
+        "bootloader" EFI_SUFFIX
+#define SYSTEMD_DST                                                                                \
+        DST_DIR                                                                                    \
+        "/"                                                                                        \
+        "loader" EFI_SUFFIX
+#define KERNEL_DST_DIR DST_DIR "/kernel"
+#define SYSTEMD_CONFIG_DIR "/loader"
+#define SYSTEMD_CONFIG SYSTEMD_CONFIG_DIR "/loader.conf"
+#define SYSTEMD_ENTRIES SYSTEMD_CONFIG_DIR "/entries"
+
+static char *shim_src;
+static char *shim_dst_host; /* as accessible by the CMB for file ops. */
+static char *shim_dst_esp;  /* absolute location of shim on the ESP. */
+static char *systemd_src;
+static char *systemd_dst_host;
+
+static char *shim_systemd_get_kernel_dst(const BootManager *manager)
+{
+        (void)manager;
+        return strdup(KERNEL_DST_DIR);
+}
+
+static bool shim_systemd_install_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        return sd_class_install_kernel_impl(manager, kernel, shim_systemd_get_kernel_dst, NULL);
+}
+
+static bool shim_systemd_remove_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        return sd_class_remove_kernel(manager, kernel);
+}
+
+static bool shim_systemd_set_default_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        /* this writes systemd config. systemd has the configuration paths
+         * hardcoded, hence whatever sd_class is doing is OK. */
+        return sd_class_set_default_kernel_impl(manager, kernel, NULL);
+}
+
+static bool exists_identical(const char *path, const char *spath)
+{
+        if (!nc_file_exists(path))
+                return false;
+        if (spath && !cbm_files_match(path, spath))
+                return false;
+        return true;
+}
+
+static bool shim_systemd_needs_install(const BootManager *manager)
+{
+        (void)manager;
+        if (!exists_identical(shim_dst_host, NULL))
+                return true;
+        if (!exists_identical(systemd_dst_host, NULL))
+                return true;
+        return false;
+}
+
+static bool shim_systemd_needs_update(const BootManager *manager)
+{
+        (void)manager;
+        if (!exists_identical(shim_dst_host, shim_src))
+                return true;
+        if (!exists_identical(systemd_dst_host, systemd_src))
+                return true;
+        return false;
+}
+
+static bool make_layout(const BootManager *manager)
+{
+        char *boot_root = boot_manager_get_boot_dir((BootManager *)manager);
+        char path[PATH_MAX];
```

Comment:
Any reason we're moving from our safe asprintf (string_printf) helpers to snprintf usage?

Reply:
the question is rather why use heap for local variables (and 3 times so)?

Reply:
So in the happy case of nothing weird went wrong with the boot_root or whatever causing it to be > PATH_MAX this code is more performant but in the unhappy case where boot_root somehow (I'm yet to track that down) is too big for PATH_MAX this snprintf will trim off characters making the path actually wrong but the mkdir may not fail which would be really nasty. I'd just want a fail safe that boot_root's size + any of the DST_DIR, KERNEL_DST_DIR and SYSTEMD_ENTRIES is under the PATH_MAX size. asprintf avoids these checks since the mkdir will just fail.

Reply:
i do not agree with this at all. it's a static aux function which uses a trusted callee. even if a call to this implementation may result in undesired side effects, the root cause can never be this implementation itself. but i will fix the code to support important use cases when `strlen(boot_var)` is comparable to addressable process memory. :)